returnn/forward.py

```diff
@@ -207,3 +216,188 @@ def hash(cls, kwargs):
+
+    def run(self):
+        """run"""
+        # run everything in a TempDir as writing files can cause heavy load
+        with tempfile.TemporaryDirectory(prefix=gs.TMP_PREFIX) as tmp_dir:
+            print("using temp-dir: %s" % tmp_dir)
+            call = [
+                self.returnn_python_exe.get_path(),
+                os.path.join(self.returnn_root.get_path(), "rnn.py"),
+                self.out_returnn_config_file.get_path(),
+            ]
+
+            try:
+                env = os.environ.copy()
+                env["OMP_NUM_THREADS"] = str(self.rqmt["cpu"])
+                env["MKL_NUM_THREADS"] = str(self.rqmt["cpu"])
+                subprocess.check_call(call, cwd=tmp_dir, env=env)
+            except Exception:
+                print("Run crashed - copy temporary work folder as 'crash_dir'")
+                if os.path.exists("crash_dir"):
+                    shutil.rmtree("crash_dir")
+                shutil.copytree(tmp_dir, "crash_dir")
+                raise
+
+            # move logs
+            if os.path.exists(f"{tmp_dir}/returnn.log"):
+                shutil.move(f"{tmp_dir}/returnn.log", "returnn.log")
+            if os.path.exists(f"{tmp_dir}/returnn-tf-log"):
+                shutil.move(f"{tmp_dir}/returnn-tf-log", ".")
+
+            # move outputs to output folder
+            for k, v in self.out_files.items():
+                assert os.path.exists(f"{tmp_dir}/{k}"), f"Output file {k} does not exist"
+                shutil.move(f"{tmp_dir}/{k}", v.get_path())
+
+    @classmethod
+    def create_returnn_config(
+        cls,
+        *,
+        model_checkpoint: Optional[Checkpoint],
+        returnn_config: ReturnnConfig,
+        log_verbosity: int,
+        device: str,
+        **_kwargs,
+    ):
+        """
+        Update the config locally to make it ready for the forward/eval task.
+        The resulting config will be used for hashing.
+
+        :param model_checkpoint:
+        :param returnn_config:
+        :param log_verbosity:
+        :param device:
+        :return:
+        """
+        assert "load" not in returnn_config.config
+        assert "model" not in returnn_config.config
+
+        res = copy.deepcopy(returnn_config)
+
+        res.config.setdefault("task", "forward")
+        if model_checkpoint is not None:
+            res.config["load"] = model_checkpoint
+        else:
+            res.config.setdefault("allow_random_model_init", True)
+
+        returnn_config.post_config.setdefault("device", device)
+        returnn_config.post_config.setdefault("log", ["./returnn.log"])
+        returnn_config.post_config.setdefault("tf_log_dir", "returnn-tf-log")
+        returnn_config.post_config.setdefault("log_verbosity", log_verbosity)
+
+        res.check_consistency()
+
+        return res
+
+    @classmethod
+    def hash(cls, kwargs):
+        d = {
+            "returnn_config": ReturnnForwardJobV2.create_returnn_config(**kwargs),
```

Comment:
Out of curiosity: Is this different from or preferred over using `cls.create_returnn_config`?

Reply:
Yes it's different in the case we would call this in a derived class. `cls` is probably better. I had it this way because I copied the `ReturnnForwardJob`.